Update example packages to use new kpt

mengqiy opened the pull request:
This PR migrates all curated functions except apply-setters and search-replace.
I will do the migration for the contrib functions in a separate PR.
I added an imperative example for each function if possible.

There are a few more examples that are planed to be added https://github.com/GoogleContainerTools/kpt/issues/1989 in a separate PR.

fixes: https://github.com/GoogleContainerTools/kpt/issues/1548

frankfarzan commented on examples/kubeval/imperative/.expected/config.yaml:
```diff
@@ -0,0 +1,8 @@
+args:
+  strict: 'true'
+stdErr:
+  - "[ERROR] Additional property templates is not allowed in object 'v1/ReplicationController//bob' in file resources.yaml in field templates"
```
With structured results design, there's a change in plan for testing. We should have a few e2e tests in the kpt repo to test the presentation layer. But otherwise, we should only test the structured output which should provide all the information we want to test the functions robustly (depend on APIs not presentation layer). In particular, packages in the catalog repo shouldn't test the presentation layer. Otherwise, you'd get brittle change detector tests:

https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html

Changing the presentation layer in the future should involve updating just a few tests.

mengqiy replied:
@frankfarzan I remember we have that discussion in another PR. Are you suggesting hold this PR until that change lands?
@Shell32-Natsu What is the status? Any ETA?

mengqiy replied:
Updated the PR to only test structured results

mengqiy replied:
>@droot is adding support GoogleContainerTools/kpt#1923

> Can you coordinate with him?

Looking at the time stamp of this comment, I guess you intended to publish last week but forgot.

mikebz commented on examples/kubeval/imperative/README.md:
````diff
@@ -0,0 +1,39 @@
+<!-- @getAndRunPkg @test -->
+```sh
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval/imperative .
+kpt fn eval imperative --image=gcr.io/kpt-fn/kubeval:unstable -- strict=true
````
very nice to show people how to do it!

mikebz replied:
I would also maybe show multiple parameters if possible.  The next question people generally have is "do I use commas or spaces?"

mengqiy replied:
I will update with multiple parameters.
Thanks for the suggestion.

mengqiy commented:
Tests related to kpt fn eval are failing because we have pick https://github.com/GoogleContainerTools/kpt/pull/1985.
I will update the eval tests

mengqiy commented:
Rebased to pick up some new fn changes.

mengqiy commented:
PTAL

phanimarupaka commented on examples/ensure-name-substring/advanced/README.md:
````diff
@@ -43,5 +58,8 @@ Check all resources have `prod-` in their names and the field `.spec.name` in
 
 ```sh
-kpt cfg cat advanced
+kpt pkg cat advanced
````
Cat is deprecated. https://kpt.dev/reference/pkg/

mengqiy replied:
Is there a replacement?
Any suggestion?

phanimarupaka replied:
linux cat specific files ?

phanimarupaka commented on examples/set-annotations/advanced/README.md:
```diff
@@ -1,14 +1,31 @@
-  ...
+  name: my-func-config
+  annotations:
+    config.kubernetes.io/local-config: 'true'
```
Is this needed in kpt v1 ?

mengqiy replied:
In kpt v1, kpt live still understand it.
I want to ensure the example package is compatible with `kpt live` (IOW the package should be able to kpt live apply after the transformation).
Otherwise, the fn config CRD will be submitted to the server and cause an error.

phanimarupaka replied:
Shouldn't kpt live also behave the same way as fn and identify such configs pointed from Kptfile ?

@frankfarzan

mengqiy replied:
+1

frankfarzan replied:
We shouldn't need `local-config` as live should understand the semantics. There's a tracking issue for this and @mortent is driving it in m3.

frankfarzan commented on .github/workflows/ci.yaml:
```diff
@@ -67,50 +67,34 @@ jobs:
+        go-version: 1.15
+    - name: Install kpt
+      run: |
+        go get github.com/GoogleContainerTools/kpt@19371100c672616db568813040de4c9adefeda3c # HEAD of next as of 05/12
```
is there an issue for this?

mengqiy replied:
Given most behaviors of render and eval have become much more stable recently, I can change to use `next`:
`go get github.com/GoogleContainerTools/kpt@next`.

frankfarzan commented on examples/apply-setters/imperative/README.md:
```diff
@@ -0,0 +1,26 @@
+
+### Overview
+
+<!-- TODO(phanimarupaka): Populate this and below -->
```
why are we adding new examples in this PR? Especially given they're incomplete.

mengqiy replied:
I can split the imperative examples out to a separate PR.

frankfarzan commented on examples/enforce-gatekeeper/imperative/README.md:
````diff
@@ -0,0 +1,88 @@
+You should see the output like the following:
+
+```
+[RUNNING] "gcr.io/kpt-fn/enforce-gatekeeper:unstable"
````
You may not want to show the non-structured output, it's likely going to change, and these dos will become stale.

frankfarzan commented on examples/enforce-gatekeeper/imperative/README.md:
````diff
@@ -0,0 +1,88 @@
+
+```shell
+$ cd imperative
+$ mkdir results
````
I would just specify `--results-dir=tmp` and you skip this step

frankfarzan commented on examples/enforce-gatekeeper/warning-only/README.md:
```diff
@@ -2,17 +2,40 @@
 ### Overview
 
-This example is very similar to the invalid configmap example. The major
+This example is very similar to the invalid configmap example. It also
```
What's "invalid configmap example"? Either link or drop the reference.

frankfarzan replied:
Please audit all the docs for go/kpt-docs-style-guide

e.g. style issue: configmap -> `ConfigMap`

frankfarzan commented on examples/enforce-gatekeeper/warning-only/README.md:
```diff
@@ -27,30 +50,37 @@ $ mkdir results
 
-You won't any failure. But if you look at the structured output, you can find a
-warning about the constraint violation.
+You won't see any failure from the console. But if you look at the structured
```
@droot why?

mengqiy replied:
Last update was 5+ days ago. The behavior may have changed already.
I will verify if this is still the latest behavior.

frankfarzan commented on examples/set-annotations/imperative/.expected/diff.patch:
```diff
@@ -0,0 +1,21 @@
+diff --git a/resources.yaml b/resources.yaml
```
General note:

We should have better file names than `resources`. Similar to naming this `configs`, it convey no information as everything is a resource, so it should be avoided.

mengqiy replied:
I'm not sure if `krm-resources.yaml` would be a better name.
Any suggested names?

frankfarzan commented on examples/set-annotations/imperative/README.md:
```diff
@@ -0,0 +1,29 @@
+# set-annotations: Imperative Example
+
+### Overview
```
nit: These should be h2 (##).

mengqiy replied:
Does this comment conflict with the 2nd item in https://github.com/GoogleContainerTools/kpt/issues/1825?

mengqiy commented:
e2e-ci is failing due to https://github.com/GoogleContainerTools/kpt/issues/2109

mengqiy commented:
The last 2 commits are new changes since last review.

Commit `update expected stderr` is caused by GoogleContainerTools/kpt#2109. I will fix it in upstream kyaml. 
The CheckSite CI issue is tracked in https://github.com/GoogleContainerTools/kpt/issues/2112 and Eyob is going to tackle it this afternoon.

frankfarzan commented on examples/ensure-name-substring/advanced/Kptfile:
```diff
@@ -0,0 +1,8 @@
+  name: example
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/ensure-name-substring:unstable
```
will unstable be removed in the final docs?

mengqiy replied:
`unstable` tag will be switched to the corresponding version tag in their release branches.
We will keep using unstable in the master branch.

frankfarzan commented on examples/kubeval/simple/README.md:
```diff
@@ -1,35 +1,65 @@
-[ERROR] Additional property templates is not allowed in object 'v1/ReplicationController//bob' in file resources.yaml in field templates
-[ERROR] Invalid type. Expected: [integer,null], given: string in object 'v1/ReplicationController//bob' in file resources.yaml in field spec.replicas
-error: exit status 1
+Package "simple": 
```
I would recommend to showing the the non-structured output which is likely to change (no backwards compatibility).

frankfarzan commented on examples/kubeval/simple/README.md:
````diff
@@ -1,35 +1,65 @@
 
 This should give the following output:
 
 ```sh
````
Style guide says uses `shell` block quote everywhere. https://github.com/GoogleContainerTools/kpt/blob/next/docs/style-guides/docs.md

frankfarzan commented on examples/set-annotations/simple/README.md:
````diff
@@ -1,35 +1,38 @@
 ```
 
-The desired annotations are provided as key-value pairs using `data` field.
+The desired annotations are provided as key-value pairs through ConfigMap.
````
https://github.com/GoogleContainerTools/kpt/blob/next/docs/style-guides/docs.md Rule 6

Please audit all the docs for style guides

frankfarzan commented on tests/e2e.sh:
```diff
@@ -17,6 +17,8 @@
 
 set -eo pipefail
-DIR="$(dirname "$0")"
+#DIR="$(dirname "$0")"
```
what's going on with this file?

mengqiy replied:
We temporarily disable a test for contrib functions, since the test depends on the old kpt.
I will find some time to fix it in a followup PR.

frankfarzan commented on tests/set-namespace/empty-input/simple/.expected/results.yaml:
```diff
@@ -1,4 +1,12 @@
-- message: 'failed to configure function: input namespace cannot be empty'
-  severity: error
+    - image: gcr.io/kpt-fn/set-namespace:unstable
+      stderr: '[error] /// : failed to configure function: function config must be a ConfigMap or SetNamespaceConfig'
```
what's going on here?

mengqiy replied:
As mentioned in my earlier comment above.
https://github.com/GoogleContainerTools/kpt/issues/2109#issuecomment-849010352 is causing this issue. I will create a PR for it in kyaml this week. It will land and be release after some time. We will bump the kyaml dependencies in functions after that. I don't think we want to block this PR because of this minor issue.

frankfarzan reviewed (approved):
remaining comments are not blocking, feel free to follow up in this PR or separate PR.

mengqiy commented:
Rebased to pick a fix for `CheckSite` CI.
The last commit contains the changes since last review.
